File: src/csharp/Core/Mesh.cs
// Mesh.cs
// Discrete Differential Geometry - Halfedge Mesh Data Structure
// Added by Graph Technologies, 2025
// Description: Complete halfedge mesh implementation for .NET

using System;
using System.Collections.Generic;
using System.Numerics;
using System.Linq;

namespace DDGCompanion.Core
{
    public class Mesh
    {
        public List<Vertex> Vertices { get; set; } = new();
        public List<Edge> Edges { get; set; } = new();
        public List<Face> Faces { get; set; } = new();
        public List<HalfEdge> HalfEdges { get; set; } = new();
        
        public void Build(Vector3[] positions, int[,] faceIndices)
        {
            if (positions == null) throw new ArgumentNullException(nameof(positions));
            if (faceIndices == null) throw new ArgumentNullException(nameof(faceIndices));
            if (faceIndices.GetLength(1) != 3)
                throw new ArgumentException("Mesh.Build currently supports triangle faces only.", nameof(faceIndices));

            Vertices.Clear();
            Edges.Clear();
            Faces.Clear();
            HalfEdges.Clear();
            
            // Create vertices
            for (int i = 0; i < positions.Length; i++)
            {
                Vertices.Add(new Vertex(positions[i]) { Index = i });
            }
            
            // Track halfedges by directed edge key and preserve endpoints.
            var halfedgeMap = new Dictionary<(int, int), HalfEdge>();
            var halfedgeEndpoints = new List<(int Source, int Target)>();
            
            // Create faces and halfedges
            for (int i = 0; i < faceIndices.GetLength(0); i++)
            {
                var face = new Face { Index = i };
                var faceHalfEdges = new List<HalfEdge>();
                
                for (int j = 0; j < 3; j++)
                {
                    int v0 = faceIndices[i, j];
                    int v1 = faceIndices[i, (j + 1) % 3];

                    if (v0 < 0 || v0 >= Vertices.Count || v1 < 0 || v1 >= Vertices.Count)
                    {
                        throw new ArgumentOutOfRangeException(nameof(faceIndices), $"Face {i} contains invalid vertex index.");
                    }
                    if (v0 == v1)
                    {
                        throw new ArgumentException($"Face {i} contains a degenerate edge ({v0} -> {v1}).", nameof(faceIndices));
                    }
                    
                    var he = new HalfEdge
                    {
                        Vertex = Vertices[v1],
                        Face = face,
                        Index = HalfEdges.Count
                    };
                    
                    faceHalfEdges.Add(he);
                    halfedgeMap[(v0, v1)] = he;
                    halfedgeEndpoints.Add((v0, v1));
                    HalfEdges.Add(he);
                }
                
                // Set next pointers
                for (int j = 0; j < 3; j++)
                {
                    faceHalfEdges[j].Next = faceHalfEdges[(j + 1) % 3];
                }
                
                face.HalfEdge = faceHalfEdges[0];
                Faces.Add(face);
            }
            
            // Set twin pointers where reverse directed edge exists.
            foreach (var (key, he) in halfedgeMap)
            {
                var (v0, v1) = key;
                var twinKey = (v1, v0);

                if (halfedgeMap.TryGetValue(twinKey, out var twin))
                {
                    he.Twin = twin;
                }
            }

            // Create undirected edges independent of winding consistency.
            var edgeMap = new Dictionary<(int, int), Edge>();
            for (int i = 0; i < HalfEdges.Count; i++)
            {
                var he = HalfEdges[i];
                var (source, target) = halfedgeEndpoints[i];
                var key = source < target ? (source, target) : (target, source);

                if (!edgeMap.TryGetValue(key, out var edge))
                {
                    edge = new Edge
                    {
                        HalfEdge = he,
                        Index = Edges.Count
                    };
                    edgeMap[key] = edge;
                    Edges.Add(edge);
                }

                he.Edge = edge;
            }
            
            // Set vertex halfedges (outgoing representative per vertex).
            for (int i = 0; i < HalfEdges.Count; i++)
            {
                var he = HalfEdges[i];
                var (source, _) = halfedgeEndpoints[i];
                if (Vertices[source].HalfEdge == null)
                {
                    Vertices[source].HalfEdge = he;
                }
            }
        }
        
        public int EulerCharacteristic() => Vertices.Count - Edges.Count + Faces.Count;
        
        public Vector3[] GetPositions()
        {
            return Vertices.Select(v => v.Position).ToArray();
        }
        
        public void SetPositions(Vector3[] positions)
        {
            for (int i = 0; i < Vertices.Count; i++)
            {
                Vertices[i].Position = positions[i];
            }
        }
        
        public void Center()
        {
            var centroid = Vector3.Zero;
            foreach (var v in Vertices)
                centroid += v.Position;
            centroid /= Vertices.Count;
            
            foreach (var v in Vertices)
                v.Position -= centroid;
        }
        
        public void Normalize()
        {
            Center();
            
            float maxDist = 0;
            foreach (var v in Vertices)
                maxDist = Math.Max(maxDist, v.Position.Length());
            
            if (maxDist > 0)
            {
                foreach (var v in Vertices)
                    v.Position /= maxDist;
            }
        }
    }
}

File: src/cpp/core/Mesh.cpp
#include "Mesh.h"
#include <map>

void Mesh::build(const Eigen::MatrixXd& V, const Eigen::MatrixXi& F) {
    vertices.clear(); edges.clear(); faces.clear(); halfedges.clear();
    
    for (int i = 0; i < V.rows(); i++) {
        vertices.push_back(std::make_unique<Vertex>(V.row(i)));
    }
    
    std::map<std::pair<int,int>, HalfEdge*> heMap;
    
    for (int i = 0; i < F.rows(); i++) {
        auto face = std::make_unique<Face>();
        std::vector<HalfEdge*> faceHEs;
        
        for (int j = 0; j < 3; j++) {
            auto he = std::make_unique<HalfEdge>();
            he->vertex = vertices[F(i, (j+1)%3)].get();
            he->face = face.get();
            faceHEs.push_back(he.get());
            heMap[{F(i,j), F(i,(j+1)%3)}] = he.get();
            halfedges.push_back(std::move(he));
        }
        
        for (int j = 0; j < 3; j++) faceHEs[j]->next = faceHEs[(j+1)%3];
        face->halfedge = faceHEs[0];
        faces.push_back(std::move(face));
    }
    
    for (auto& [key, he] : heMap) {
        auto twin = heMap.find({key.second, key.first});
        if (twin != heMap.end()) {
            he->twin = twin->second;
            if (!he->edge) {
                auto edge = std::make_unique<Edge>();
                edge->halfedge = he;
                he->edge = edge.get();
                twin->second->edge = edge.get();
                edges.push_back(std::move(edge));
            }
        }
    }
    
    for (auto& he : halfedges) {
        if (!he->vertex->halfedge) he->vertex->halfedge = he.get();
    }
    reindex();
}

void Mesh::reindex() {
    for (size_t i = 0; i < vertices.size(); i++) vertices[i]->index = i;
    for (size_t i = 0; i < edges.size(); i++) edges[i]->index = i;
    for (size_t i = 0; i < faces.size(); i++) faces[i]->index = i;
}

Eigen::MatrixXd Mesh::vertexPositions() const {
    Eigen::MatrixXd V(numVertices(), 3);
    for (size_t i = 0; i < numVertices(); i++) V.row(i) = vertices[i]->position;
    return V;
}

void Mesh::setVertexPositions(const Eigen::MatrixXd& V) {
    for (size_t i = 0; i < numVertices(); i++) vertices[i]->position = V.row(i);
}

void Mesh::center() {
    Eigen::Vector3d c = Eigen::Vector3d::Zero();
    for (const auto& v : vertices) c += v->position;
    c /= numVertices();
    for (auto& v : vertices) v->position -= c;
}

void Mesh::normalize() {
    center();
    double maxD = 0;
    for (const auto& v : vertices) maxD = std::max(maxD, v->position.norm());
    if (maxD > 0) for (auto& v : vertices) v->position /= maxD;
}
